awssql/internal_pool/internal_pooled_connection_provider.go

```diff
@@ -0,0 +1,129 @@
+/*
+  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License").
+  You may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package internal_pool
+
+import (
+	"database/sql/driver"
+	"time"
+
+	"github.com/aws/aws-advanced-go-wrapper/awssql/driver_infrastructure"
+	"github.com/aws/aws-advanced-go-wrapper/awssql/error_util"
+	"github.com/aws/aws-advanced-go-wrapper/awssql/host_info_util"
+	"github.com/aws/aws-advanced-go-wrapper/awssql/property_util"
+	"github.com/aws/aws-advanced-go-wrapper/awssql/utils"
+)
+
+type internalPoolKeyFunc func(*host_info_util.HostInfo, map[string]string) string
+
+type InternalPooledConnectionProvider struct {
+	acceptedStrategies     map[string]driver_infrastructure.HostSelector
+	targetDriver           driver.Driver
+	databasePools          *utils.SlidingExpirationCache[*InternalConnPool]
+	internalPoolOptions    *InternalPoolConfig
+	poolKeyFunc            internalPoolKeyFunc
+	poolExpirationDuration time.Duration
+}
+
+func NewInternalPooledConnectionProvider(targetDriver driver.Driver,
```

Comment:
Going to try to remove the `targetDriver` parameter and see if I can make this automatic.